databuilder/population_validation.py

```diff
@@ -0,0 +1,225 @@
+    # with a positively defined population (e.g. all patients registered with a practice
+    # as of a certain date) from which some patients are then excluded. A population
+    # definition like "all patients who have not died" doesn't unambiguously define a
+    # universe of patients and so should not be accepted.
+    if evaluate(population) is True:
+        # TODO: Wording could do with more thought here
+        raise PopulationValidationError(
+            "population definition must not evaluate as True for NULL inputs"
+        )
+    return True
+
+
+@singledispatch
+def evaluate(node):
+    """
+    Given a Series node, return the value it would take if all its inputs were NULL
+    """
+    assert False, f"Unhandled node type: {type(node)}"
+
+
+@evaluate.register(Value)
+def evaluate_value(node):
+    return node.value
+
+
+@evaluate.register(SelectColumn)
+def evaluate_select_column(node):
+    return None
+
+
+@evaluate.register(AggregateByPatient.Exists)
+def evaluate_exists(node):
+    return False
+
+
+@evaluate.register(AggregateByPatient.Count)
+def evaluate_count(node):
+    return 0
+
+
+@evaluate.register(AggregateByPatient.CombineAsSet)
+def evaluate_combine_as_set(node):
+    return ()
+
+
+@evaluate.register(AggregateByPatient.Min)
+@evaluate.register(AggregateByPatient.Max)
+@evaluate.register(AggregateByPatient.Sum)
+def evaluate_aggregation(node):
+    return None
+
+
+@evaluate.register(Function.IsNull)
+def evaluate_is_null(node):
+    return evaluate(node.source) is None
+
+
+@evaluate.register(Function.And)
+def evaluate_and(node):
+    lhs = evaluate(node.lhs)
+    rhs = evaluate(node.rhs)
+    if lhs is True and rhs is True:
+        return True
+    elif lhs is False or rhs is False:
+        return False
+    else:
+        return None
+
+
+@evaluate.register(Function.Or)
+def evaluate_or(node):
+    lhs = evaluate(node.lhs)
+    rhs = evaluate(node.rhs)
+    if lhs is True or rhs is True:
+        return True
+    elif lhs is False and rhs is False:
+        return False
+    else:
+        return None
+
+
+@evaluate.register(Categorise)
+def evaluate_categorise(node):
+    for value, condition in node.categories.items():
+        if evaluate(condition) is True:
+            return evaluate(value)
+    if node.default is not None:
+        return evaluate(node.default)
+
+
+def register_op(cls):
+    """
+    Handle registration for any operations with the default NULL behaviour, which is
+    that they return NULL if any of their arguments are NULL
+    """
+    getters = [operator.attrgetter(field.name) for field in dataclasses.fields(cls)]
+
+    def register(function):
+        @evaluate.register(cls)
+        def apply(node):
+            args = [evaluate(get_arg(node)) for get_arg in getters]
+            if any(arg is None for arg in args):
+                return None
+            return function(*args)
+
+        return function
```

Comment:
Sort of yes and sort of no :) It becomes the definition of the decorated function in the module namespace. As it happens we never call them directly so they are "thrown away" in that sense. But it's a pretty standard pattern for registration decorators and I'd feel a bit weird about (implicitly) returning None here.

Reply:
Oh I see, yes. I was trying to follow the pattern of `singledispatch.register` which returns the original function unchanged. But `@register_op` is a weird hybrid decorator that both modifies the function and registers it as an `evaluate` implementation. It would probably be cleaner to separate those roles out, but it would be slightly more typing. Returning `apply` seems reasonable if that's less surprising.